### PR TITLE
[CORE] Add RemoveKnownFloatingPointNormalized Rule

### PR DESCRIPTION
**Pull request by liujiayi771**
## What changes were proposed in this pull request?

We plan to pull out the pre-project of agg in the logical plan. However, due to some historical reasons, Spark's agg `groupingExpressions` will add `KnownFloatingPointNormalized` to certain specific data types when converting to a physical plan. These expressions need to be pull out into the pre-project during native execution, but we have not pulled out the pre-project of agg in the physical plan. `KnownFloatingPointNormalized` does not actually serve any purpose in Gluten, and during plan conversion, we directly processes its child, so we can remove it.

## How was this patch tested?

ClickHouse backend already have a "knownfloatingpointnormalized" test case in `GlutenClickHouseFileFormatSuite`. Add a "knownfloatingpointnormalized" test case in Velox backend `TestOperator`.

### CODE REVIEW

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 1]
```diff
@@ -786,6 +786,7 @@ case class ColumnarOverrideRules(session: SparkSession)
       List(
         (spark: SparkSession) => PlanOneRowRelation(spark),
         (_: SparkSession) => FallbackEmptySchemaRelation(),
+        (spark: SparkSession) => RemoveKnownFloatingPointNormalized(spark),
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:789]: We should apply this rule after `AddTransformHintRule` and check that we apply this rule only if we support transform aggregate

File: gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+  override def apply(plan: SparkPlan): SparkPlan = PhysicalPlanSelector.maybe(session, plan) {
+    plan.transform {
+      case agg: BaseAggregateExec
+          if normalizeExpressionExists(agg.groupingExpressions) &&
+            !TransformHints.isAlreadyTagged(plan) =>
+        val newGroupingExprs = agg.groupingExpressions.toIndexedSeq.map {
+          expr =>
+            val transformedExpr = expr.transform {
+              case _ @KnownFloatingPointNormalized(_ @NormalizeNaNAndZero(child)) => child
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala:60]: Can we remove `KnownFloatingPointNormalized` and `NormalizeNaNAndZero` in `ExpressionConverter` ?

Comment (liujiayi771) [reply gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala]: Yes. Removed.

Comment (liujiayi771) [reply gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala:63]: @ulysses-you We cannot remove, both window and join still use them. They are handled in the logical plan rule `NormalizeFloatingNumbers`, so we don't need to eliminate it. We only need to remove the `NormalizeFloatingNumbers` in agg. Although we could also include join and window in `RemoveKnownFloatingPointNormalized`, I think it's unnecessary. In the future, when Spark handles agg within the logical plan as well, we can remove `RemoveKnownFloatingPointNormalized` in Gluten.

Comment (ulysses-you) [reply gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala:63]: I see, after we pull out join/window, the NormalizeNaNAndZero will be moved to project and we skip tranform it at the end

File: gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      case agg: BaseAggregateExec
+          if normalizeExpressionExists(agg.groupingExpressions) &&
+            (validation || supportTransform(plan)) =>
+        val newGroupingExprs = agg.groupingExpressions.toIndexedSeq.map {
+          expr =>
+            val transformedExpr = expr.transform {
+              case _ @KnownFloatingPointNormalized(_ @NormalizeNaNAndZero(child)) => child
+              case _ @KnownFloatingPointNormalized(_ @If(_ @IsNull(expr), _, _)) => expr
+              case _ @KnownFloatingPointNormalized(_ @ArrayTransform(expr, _)) => expr
+              case other => other
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala:67]: transform do not need other pattern

File: gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    expressions.exists(_.find(_.isInstanceOf[KnownFloatingPointNormalized]).isDefined)
+  }
+
+  /**
+   * If the groupingExpressions of the aggregate are expressions, an Alias will be added on the
+   * outside. After removing the KnownFloatingPointNormalized, it is necessary to remove this
+   * redundant Alias.
+   */
+  private def removeRedundantAlias(expression: Expression): Expression = expression match {
+    case _ @Alias(child: Attribute, _) => child
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala:46]: `_ @` is unused

File: gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      // Only remove KnownFloatingPointNormalized when agg is transformable.
+      case agg: BaseAggregateExec
+          if normalizeExpressionExists(agg.groupingExpressions) &&
+            (validation || supportTransform(plan)) =>
+        val newGroupingExprs = agg.groupingExpressions.toIndexedSeq.map {
+          expr =>
+            val transformedExpr = expr.transform {
+              case _ @KnownFloatingPointNormalized(_ @NormalizeNaNAndZero(child)) => child
+              case _ @KnownFloatingPointNormalized(_ @If(_ @IsNull(expr), _, _)) => expr
+              case _ @KnownFloatingPointNormalized(_ @ArrayTransform(expr, _)) => expr
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala:66]: `_ @` is unused

File: gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    TransformHints.isAlreadyTagged(plan) && TransformHints.isTransformable(plan)
+
+  override def apply(plan: SparkPlan): SparkPlan = applyWithValidation(plan)
+
+  def applyWithValidation(plan: SparkPlan, validation: Boolean = false): SparkPlan =
+    plan.transform {
+      // Only remove KnownFloatingPointNormalized when agg is transformable.
+      case agg: BaseAggregateExec
+          if normalizeExpressionExists(agg.groupingExpressions) &&
+            (validation || supportTransform(plan)) =>
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala:60]: `supportTransform(agg)`

File: gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        val newGroupingExprs = agg.groupingExpressions.toIndexedSeq.map {
+          expr =>
+            val transformedExpr = expr.transform {
+              case _ @KnownFloatingPointNormalized(_ @NormalizeNaNAndZero(child)) => child
+              case _ @KnownFloatingPointNormalized(_ @If(_ @IsNull(expr), _, _)) => expr
+              case _ @KnownFloatingPointNormalized(_ @ArrayTransform(expr, _)) => expr
+              case other => other
+            }
+            removeRedundantAlias(transformedExpr)
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala:69]: we can do `. asInstanceOf[NamedExpression]` here

**Comment by liujiayi771** [comment]
@ulysses-you Aside from the situation in this PR, it is also possible to insert agg into the physical plan within the DPP. Therefore, I am planning to perform an additional pre-project check in the physical plan. In most cases, pre-project can be inserted in the logical plan, and any omissions can be compensated for in the physical plan. This approach can also preserve the performance advantages of processing in the logical plan. In that case, this PR would no longer be necessary.

File: gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            val transformedExpr = expr.transform {
+              case KnownFloatingPointNormalized(NormalizeNaNAndZero(child)) => child
+              case KnownFloatingPointNormalized(If(IsNull(expr), _, _)) => expr
+              case KnownFloatingPointNormalized(ArrayTransform(expr, _)) => expr
+            }
+            removeRedundantAlias(transformedExpr).asInstanceOf[NamedExpression]
+        }
+        agg match {
+          case hash: HashAggregateExec =>
+            hash.copy(groupingExpressions = newGroupingExprs)
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/extension/RemoveKnownFloatingPointNormalized.scala:72]: It seems we should fail in `TransformPreOverrides` as  we apply `RemoveKnownFloatingPointNormalized` first and do not copy tags, but the tests passed.. can you check if this rule is actually applied ?
https://github.com/oap-project/gluten/blob/451b1cee89b9425443924ae69f9dd04a186c85a4/gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala#L282C1-L283

**Comment by ulysses-you** [comment]
@liujiayi771 dpp should not be affected as it is a subquery. This new rule should be applied in it's own AQE.

**Comment by liujiayi771** [comment]
> @liujiayi771 dpp should not be affected as it is a subquery. This new rule should be applied in it's own AQE.

I find a unit test error with ClickHouse. Upon debugging, I noticed that agg is inserted at a physical rule. So, the logical plan rule wouldn't be able to see this agg, right?

https://github.com/apache/spark/blob/86bac8911a31a98eb6ad1f5365554f4f095c0376/sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PlanDynamicPruningFilters.scala#L84C19-L84C19

**Pull request closed without merging**